File: legacy-python/agents/player_agent.py
from collections import deque
from agents.dao_member import DAOMember
from utils.proposal_utils import create_random_proposal


class PlayerAgent(DAOMember):
    """Agent controlled via external API calls."""

    def __init__(self, unique_id, model, tokens, reputation, location):
        super().__init__(unique_id, model, tokens, reputation, location)
        self.actions = deque()

    # ------------------------------------------------------------------
    # Helper methods interacting with the DAO treasury

    def _treasury_swap(self, token_in: str, token_out: str, amount: float) -> float:
        treasury = self.model.treasury
        step = getattr(self.model, "current_step", 0)
        treasury.deposit(token_in, amount, step=step)
        try:
            out = treasury.swap(token_in, token_out, amount, step=step)
        except ValueError:
            treasury.withdraw(token_in, amount, step=step)
            return 0.0
        gained = treasury.withdraw(token_out, out, step=step)
        return gained

    def _treasury_add_liquidity(
        self, token_a: str, token_b: str, amt_a: float, amt_b: float
    ) -> None:
        treasury = self.model.treasury
        step = getattr(self.model, "current_step", 0)
        treasury.deposit(token_a, amt_a, step=step)
        treasury.deposit(token_b, amt_b, step=step)
        treasury.add_liquidity(token_a, token_b, amt_a, amt_b, step=step)

    def _treasury_remove_liquidity(
        self, token_a: str, token_b: str, share: float
    ) -> tuple[float, float]:
        treasury = self.model.treasury
        step = getattr(self.model, "current_step", 0)
        amt_a, amt_b = treasury.remove_liquidity(token_a, token_b, share, step=step)
        gained_a = treasury.withdraw(token_a, amt_a, step=step)
        gained_b = treasury.withdraw(token_b, amt_b, step=step)
        return gained_a, gained_b

    def enqueue(self, action: str, **kwargs) -> None:
        """Add an action to the queue."""
        self.actions.append((action, kwargs))

    def step(self) -> None:
        if not self.actions:
            return
        action, kwargs = self.actions.popleft()
        if action == "vote":
            proposal = kwargs.get("proposal")
            vote = kwargs.get("vote", True)
            if proposal is not None:
                proposal.add_vote(self, vote)
                self.votes[proposal] = {"vote": vote, "weight": 1}
                self.mark_active()
        elif action == "comment":
            proposal = kwargs.get("proposal")
            sentiment = kwargs.get("sentiment", "neutral")
            if proposal is not None:
                self.leave_comment(proposal, sentiment)
        elif action == "create_proposal":
            proposal = create_random_proposal(self.model, self)
            self.model.add_proposal(proposal)
            self.mark_active()
        elif action == "delegate":
            proposal = kwargs.get("proposal")
            amount = kwargs.get("amount", 0)
            if proposal is not None and amount > 0 and self.tokens >= amount:
                self.tokens -= amount
                proposal.receive_delegated_support(self, amount)
                self.mark_active()
        elif action == "swap":
            token_in = kwargs.get("token_in")
            token_out = kwargs.get("token_out")
            amount = kwargs.get("amount", 0.0)
            if token_in and token_out and amount > 0 and self.tokens >= amount:
                gained = self._treasury_swap(token_in, token_out, amount)
                self.tokens -= amount
                self.tokens += gained
                self.mark_active()
        elif action == "add_liquidity":
            token_a = kwargs.get("token_a")
            token_b = kwargs.get("token_b")
            amt_a = kwargs.get("amount_a", 0.0)
            amt_b = kwargs.get("amount_b", 0.0)
            total = amt_a + amt_b
            if token_a and token_b and total > 0 and self.tokens >= total:
                self._treasury_add_liquidity(token_a, token_b, amt_a, amt_b)
                self.tokens -= total
                self.mark_active()
        elif action == "remove_liquidity":
            token_a = kwargs.get("token_a")
            token_b = kwargs.get("token_b")
            share = kwargs.get("share", 0.0)
            if token_a and token_b and 0 < share <= 1:
                amt_a, amt_b = self._treasury_remove_liquidity(token_a, token_b, share)
                self.tokens += amt_a + amt_b
                self.mark_active()
        else:
            # unknown action - ignore
            pass

File: lib/agents/player-agent.ts
// Player Agent - controlled via external API calls
// Port from agents/player_agent.py

import { DAOMember } from './base';
import type { DAOModel } from '../engine/model';
import type { Proposal } from '../data-structures/proposal';
import { createRandomProposal } from '../utils/proposal-utils';

type PlayerAction =
  | { type: 'vote'; proposal: Proposal; vote: boolean }
  | { type: 'comment'; proposal: Proposal; sentiment: string }
  | { type: 'create_proposal' }
  | { type: 'delegate'; proposal: Proposal; amount: number }
  | { type: 'swap'; tokenIn: string; tokenOut: string; amount: number }
  | { type: 'add_liquidity'; tokenA: string; tokenB: string; amountA: number; amountB: number }
  | { type: 'remove_liquidity'; tokenA: string; tokenB: string; share: number };

export class PlayerAgent extends DAOMember {
  actions: PlayerAction[] = [];

  constructor(
    uniqueId: string,
    model: DAOModel,
    tokens: number = 100,
    reputation: number = 50,
    location: string = 'node_0'
  ) {
    super(uniqueId, model, tokens, reputation, location);
  }

  /**
   * Helper: Swap tokens via treasury
   */
  private treasurySwap(tokenIn: string, tokenOut: string, amount: number): number {
    const treasury = this.model.dao.treasury;
    const step = this.model.currentStep;

    treasury.deposit(tokenIn, amount, step);
    try {
      const out = treasury.swap(tokenIn, tokenOut, amount, step);
      const gained = treasury.withdraw(tokenOut, out, step);
      return gained;
    } catch {
      // Swap failed, refund
      treasury.withdraw(tokenIn, amount, step);
      return 0;
    }
  }

  /**
   * Helper: Add liquidity to pool
   */
  private treasuryAddLiquidity(
    tokenA: string,
    tokenB: string,
    amtA: number,
    amtB: number
  ): void {
    const treasury = this.model.dao.treasury;
    const step = this.model.currentStep;

    treasury.deposit(tokenA, amtA, step);
    treasury.deposit(tokenB, amtB, step);
    treasury.addLiquidity(tokenA, tokenB, amtA, amtB, step);
  }

  /**
   * Helper: Remove liquidity from pool
   */
  private treasuryRemoveLiquidity(
    tokenA: string,
    tokenB: string,
    share: number
  ): [number, number] {
    const treasury = this.model.dao.treasury;
    const step = this.model.currentStep;

    const [amtA, amtB] = treasury.removeLiquidity(tokenA, tokenB, share, step);
    const gainedA = treasury.withdraw(tokenA, amtA, step);
    const gainedB = treasury.withdraw(tokenB, amtB, step);

    return [gainedA, gainedB];
  }

  /**
   * Enqueue an action to be executed in the next step
   */
  enqueue(action: PlayerAction): void {
    this.actions.push(action);
  }

  step(): void {
    if (this.actions.length === 0) {
      return;
    }

    const action = this.actions.shift()!;

    switch (action.type) {
      case 'vote':
        action.proposal.addVote(this.uniqueId, action.vote, 1);
        this.votes.set(action.proposal.uniqueId, {
          vote: action.vote,
          weight: 1,
        });
        this.markActive();
        break;

      case 'comment':
        this.leaveComment(action.proposal, action.sentiment);
        break;

      case 'create_proposal': {
        const proposal = createRandomProposal(this.model.dao, this);
        this.model.dao.addProposal(proposal);
        this.markActive();
        break;
      }

      case 'delegate':
        if (action.amount > 0 && this.tokens >= action.amount) {
          this.tokens -= action.amount;
          (action.proposal as any).receiveDelegatedSupport?.(
            this.uniqueId,
            action.amount
          );
          this.markActive();
        }
        break;

      case 'swap':
        if (action.amount > 0 && this.tokens >= action.amount) {
          const gained = this.treasurySwap(
            action.tokenIn,
            action.tokenOut,
            action.amount
          );
          this.tokens -= action.amount;
          this.tokens += gained;
          this.markActive();
        }
        break;

      case 'add_liquidity': {
        const total = action.amountA + action.amountB;
        if (total > 0 && this.tokens >= total) {
          this.treasuryAddLiquidity(
            action.tokenA,
            action.tokenB,
            action.amountA,
            action.amountB
          );
          this.tokens -= total;
          this.markActive();
        }
        break;
      }

      case 'remove_liquidity':
        if (action.share > 0 && action.share <= 1) {
          const [amtA, amtB] = this.treasuryRemoveLiquidity(
            action.tokenA,
            action.tokenB,
            action.share
          );
          this.tokens += amtA + amtB;
          this.markActive();
        }
        break;
    }
  }
}
